static/js/emoji.js

```diff
@@ -124,13 +124,35 @@ exports.get_canonical_name = function (emoji_name) {
 // Translates emoticons in a string to their colon syntax.
 exports.translate_emoticons_to_names = function translate_emoticons_to_names(text) {
     var translated = text;
+    var replacement_text;
+    var terminal_symbols = ',.;?!()[] "\'\n\t'; // From composebox_typeahead
+    var symbols_except_space = terminal_symbols.replace(' ', '');
+
+    var emoticon_replacer = function (match, g1, offset, str) {
+        var prev_char = str[offset - 1];
+        var next_char = str[offset + match.length];
+
+        var symbol_at_start = terminal_symbols.indexOf(prev_char) !== -1;
+        var symbol_at_end = terminal_symbols.indexOf(next_char) !== -1;
+        var non_space_at_start = symbols_except_space.indexOf(prev_char) !== -1;
+        var non_space_at_end = symbols_except_space.indexOf(next_char) !== -1;
+        var valid_start = symbol_at_start || offset === 0;
+        var valid_end = symbol_at_end || offset === str.length - match.length;
+
+        if (non_space_at_start && non_space_at_end) { // Hello!:)?
+            return match;
+        }
+        if (valid_start && valid_end) {
+            return replacement_text;
+        }
+        return match;
```

Comment:
The logic seems like it could be simplified here, into maybe

```js
if ((valid_start && valid_end) && !(non_space_at_start && non_space_at_end)) {
    return replacement_text;
}

return match;
```

instead of having both an `if` statement and the default cause going towards the same result.

Reply:
You're correct. I was deciding between readability and good flow of control. I'll change it after classes get over..

Reply:
I think readability is most important; I think doing this + adding a bit of a block comment explaining what we're trying to do and how the motivation is driven in part by the lack of negative lookaheads would be valuable here.